### DISCUSSION TITLE
How to detect multi-line comments without closing?

### DISCUSSION THREAD

**Discussion by durianwaffle** [eclipse-langium/langium#1719] [Q&A] [answered]
Hi Langium community,

I am using comment token to check if current cursor is in multi-line comment.
But the problem is if the comment is not closed, it will not have the comment token.

May I know if there are existing services to detect if current cursor location is in multi-line comment?

**Comment by wa8f7192** [comment]
Hi,

To addon to the comment, attaching examples here

Works for these examples
```
/*
<user's cursor>
*/
```

```
/* <user's cursor>
*/
```

```
// Single line comment works as well <user's cursor>
```

Doesn't work for this
```
/* <user's cursor>
```

Our current code
```
// Code to detect if the cursor position is at a comment
export const positionIsAtComment = (document: LangiumDocument<AstNode>, position: Position): boolean => {
    const documentCstNode = document.parseResult.value.$cstNode;
    if (!documentCstNode) {
        return false;
    }
    const offset = document.textDocument.offsetAt(position);
    const cstNode = CstUtils.findDeclarationNodeAtOffset(documentCstNode, offset);
    

    if (cstNode?.tokenType) {
        return cstNode.tokenType.name.toUpperCase().includes('COMMENT');
    }
    return false;
}
```

These are the rules used for terminal for us
```
hidden terminal WS: /\s+/;
terminal ID: /[_a-zA-Z][\w_]*/;
terminal NUMBER returns number: /[0-9]+(\.[0-9]+)?([eE][+-]?[0-9]+)?/;
hidden terminal ML_COMMENT: /(\/\*[\s\S]*?\*\/)/;
hidden terminal ML_COMMENT1: '(*' -> '*)';
hidden terminal SL_COMMENT: /\/\/[^\n\r]*/;
```

**Comment by msujew** [comment, marked as answer]
Hey @durianwaffle, @wa8f7192,

you should be able to do something like that:

```
hidden terminal ML_COMMENT_INCOMPLETE: /(\/\*[\s\S]*)/;
```

and move that behind the existing `ML_COMMENT` terminal. That way you don't accidentally hide the existing comment terminal.

**Reply by durianwaffle** [reply]
Thanks a lot! It works!

**Reply by durianwaffle** [reply]
But I found some issues for nested comments:
![image](https://github.com/user-attachments/assets/76c30522-3593-41c1-bcf0-2e18468b2c64)

![image](https://github.com/user-attachments/assets/82233369-81ab-4eae-81be-d4025444f53d)

![image](https://github.com/user-attachments/assets/54a1d607-b0c4-4e44-82f0-f78e1b429fb2)

**Reply by msujew** [reply]
You cannot create nested comments with regex, as that would require recursion (or at least a stack). You need to write the `PATTERN` function for the token type yourself. See the `TokenBuilder` for more info.
